File: clojure/day24_part1_2023.clj

(defn parse-line [line]
  (let [[px py pz vx vy vz] (map #(Double/parseDouble %) (re-seq #"-?\d+\.?\d*" line))]
    {:pos {:x px :y py :z pz} :vel {:x vx :y vy :z vz}}))

(defn parse-input [input]
  (map parse-line input))

(defn intersection-2d [p1 p2]
  (let [det (- (* (:x (:vel p1)) (:y (:vel p2))) (* (:x (:vel p2)) (:y (:vel p1))))]
    (if (zero? det)
      nil
      (let [t1 (/ (- (* (:y (:vel p2)) (- (:x (:pos p2)) (:x (:pos p1)))) (* (:x (:vel p2)) (- (:y (:pos p2)) (:y (:pos p1))))) det)
            t2 (/ (- (* (:y (:vel p1)) (- (:x (:pos p2)) (:x (:pos p1)))) (* (:x (:vel p1)) (- (:y (:pos p2)) (:y (:pos p1))))) det)
            x (+ (:x (:pos p1)) (* (:x (:vel p1)) t1))
            y (+ (:y (:pos p1)) (* (:y (:vel p1)) t1))]
        {:coord {:x x :y y} :t1 t1 :t2 t2}))))

(defn solve [input min-val max-val]
  (let [points (parse-input input)]
    (->> (for [i (range (count points))
               j (range i)]
           (let [p1 (nth points i)
                 p2 (nth points j)
                 intersection (intersection-2d p1 p2)]
             (when intersection
               (let [{:keys [coord t1 t2]} intersection
                     {:keys [x y]} coord]
                 (and (>= x min-val) (<= x max-val)
                      (>= y min-val) (<= y max-val)
                      (>= t1 0) (>= t2 0))))))
         (filter identity)
         count)))

(defn read-file [file-name]
  (-> (slurp file-name)
      (clojure.string/trim)
      (clojure.string/split #"\n")))

(defn -main []
  (let [input (read-file "input.txt")]
    (println (solve input 200000000000000.0 400000000000000.0))))
(-main)

File: fsharp/day24_part1_2023.fs

open System
open System.IO

type Point = { Px:double; Py:double; Pz:double; Vx:double; Vy:double; Vz:double }

let parse (s:string) =
    let p = s.Split([|'@'|], StringSplitOptions.RemoveEmptyEntries)
    let pos = p.[0].Split([|','|], StringSplitOptions.RemoveEmptyEntries) |> Array.map (fun x -> Double.Parse(x.Trim()))
    let vel = p.[1].Split([|','|], StringSplitOptions.RemoveEmptyEntries) |> Array.map (fun x -> Double.Parse(x.Trim()))
    { Px=pos.[0]; Py=pos.[1]; Pz=pos.[2]; Vx=vel.[0]; Vy=vel.[1]; Vz=vel.[2] }

let intersects p1 p2 =
    let det = p1.Vx * p2.Vy - p2.Vx * p1.Vy
    if abs det < 1e-12 then None else
    let dt = (p2.Px - p1.Px, p2.Py - p1.Py)
    let t1 = (p2.Vy * fst dt - p2.Vx * snd dt) / det
    let t2 = (p1.Vy * fst dt - p1.Vx * snd dt) / det
    if t1 < 0.0 || t2 < 0.0 then None else
    let x = p1.Px + p1.Vx * t1
    let y = p1.Py + p1.Vy * t1
    Some (x, y)

[<EntryPoint>]
let main _ =
    let pts = File.ReadAllLines("input.txt") |> Array.map parse
    let lo, hi = 200000000000000.0, 400000000000000.0
    let ok p =
        let x, y = p
        x >= lo && x <= hi && y >= lo && y <= hi
    seq { for i in 0..pts.Length-1 do
          for j in 0..i-1 do
          match intersects pts.[i] pts.[j] with
          | Some p when ok p -> yield () | _ -> () }
    |> Seq.length |> printfn "%d"
    0
